File: android-private/fitai-android/wear/src/main/java/com/fitai/wear/WatchTimer.kt
package com.fitai.wear

import android.os.SystemClock
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.max

fun createRestDeadline(restRemaining: Int, elapsedNowMs: Long = SystemClock.elapsedRealtime()): RestDeadline? {
    val seconds = max(0, restRemaining)
    if (seconds <= 0) return null
    return RestDeadline(
        deadlineElapsedMs = elapsedNowMs + seconds * 1_000L,
        sourceRemainingSeconds = seconds,
    )
}

fun remainingFromDeadline(deadline: RestDeadline?, elapsedNowMs: Long = SystemClock.elapsedRealtime()): Int {
    if (deadline == null) return 0
    return max(0, ceil((deadline.deadlineElapsedMs - elapsedNowMs) / 1_000.0).toInt())
}

fun shouldReplaceDeadline(
    current: RestDeadline?,
    next: RestDeadline?,
    contextChanged: Boolean,
    elapsedNowMs: Long = SystemClock.elapsedRealtime(),
): Boolean {
    if (contextChanged) return true
    if (current == null || next == null) return current != next
    val currentRemaining = remainingFromDeadline(current, elapsedNowMs)
    val nextRemaining = remainingFromDeadline(next, elapsedNowMs)
    return abs(currentRemaining - nextRemaining) > 2
}

fun formatRest(seconds: Int): String {
    if (seconds <= 0) return "GO"
    if (seconds < 60) return seconds.toString()
    val minutes = seconds / 60
    val rest = seconds % 60
    return "$minutes:${rest.toString().padStart(2, '0')}"
}

File: src/lib/watch-timer.ts
export type WatchRestDeadline = {
  endsAtEpochMs: number;
  receivedAtEpochMs: number;
  receivedAtPerfMs: number;
};

export function formatWatchRest(seconds: number) {
  const safeSeconds = Math.max(0, Math.ceil(seconds));
  if (safeSeconds <= 0) return "GO";
  if (safeSeconds < 60) return String(safeSeconds);

  const minutes = Math.floor(safeSeconds / 60);
  const remainder = safeSeconds % 60;
  return `${minutes}:${String(remainder).padStart(2, "0")}`;
}

export function getRemainingFromDeadline(deadline: WatchRestDeadline | null, perfNowMs: number) {
  if (!deadline) return 0;

  const elapsedMs = Math.max(0, perfNowMs - deadline.receivedAtPerfMs);
  const remainingMs = deadline.endsAtEpochMs - deadline.receivedAtEpochMs - elapsedMs;
  return Math.max(0, Math.ceil(remainingMs / 1000));
}

export function createRestDeadlineFromServer(input: {
  restRemaining: number;
  receivedAtEpochMs: number;
  receivedAtPerfMs: number;
}) {
  const remainingSeconds = Math.max(0, Math.floor(input.restRemaining));
  if (remainingSeconds <= 0) return null;

  return {
    endsAtEpochMs: input.receivedAtEpochMs + remainingSeconds * 1000,
    receivedAtEpochMs: input.receivedAtEpochMs,
    receivedAtPerfMs: input.receivedAtPerfMs,
  };
}

export function shouldReplaceRestDeadline(input: {
  current: WatchRestDeadline | null;
  next: WatchRestDeadline | null;
  perfNowMs: number;
  contextChanged: boolean;
  toleranceSeconds?: number;
}) {
  if (input.contextChanged) return true;
  if (!input.current || !input.next) return input.current !== input.next;

  const toleranceSeconds = input.toleranceSeconds ?? 1.25;
  const currentRemaining = getRemainingFromDeadline(input.current, input.perfNowMs);
  const nextRemaining = getRemainingFromDeadline(input.next, input.next.receivedAtPerfMs);
  return Math.abs(currentRemaining - nextRemaining) > toleranceSeconds;
}
